docker/cassandra/Dockerfile

```diff
@@ -10,6 +10,13 @@ COPY install-cassandra /tmp/install-cassandra
 RUN /tmp/install-cassandra ${cassandra_version}
 
 COPY entry-point /usr/bin/entry-point
+COPY wrapper /usr/bin/wrapper
+
+# Allow entrypoint script to modify ulimit by creating a "patched" Bash executable
+# which is allowed to use the required capabilities
+RUN cp /bin/bash /bin/bash-mod \
```

Comment:
This seems a little odd to me. Why are we doing ulimit modifications via a "patched" bash?

Reply:
@benbromhead I think this solution originates from here (1) and distilled comments bellow it. Our "exe" is basically `bash` which is executing entry-point from wrapper script where Cassandra is started.

By using setcap below in this file, that "exe" (patched bash) can set ulimits (without root) which is what we want.

@johananl is that how it works for sure? Is my understanding correct?

(1) https://github.com/kubernetes/kubernetes/issues/3595#issuecomment-438507708

Reply:
@benbromhead what @smiklosovic wrote is correct. I've tried to explain the rationale in the description but let me recap here:

- We want to let the C* process lock memory while running as a non-root user.
- We can't give C* the required Linux capabilities via k8s, which means we have to do it somehow from within the container.
- We want the above to be *conditional*, i.e. we should be able to also run without requiring the extra capabilities, for example when running on clusters with restrictive security policies.
- C* runs a Bash script on startup which tries to do `ulimit -l unlimited`. The **binary** which does so (which is `/bin/bash`) needs to have the `SYS_RESOURCE` capability. Additionally, C* needs to use `mlock` syscalls. For that, the **binary** used for spawning the C* process needs to have the `IPC_LOCK` capability. This binary would normally be the JVM, however since we `exec` the C* process from the entrypoint script (rather than creating a child process), we can set the `IPC_LOCK` capability on the Bash binary as well.
- The reason we need to have both a Bash binary with and without the capabilities is that when we run in "low privileged" mode (i.e. without these tweaks), the C* container won't start if `/bin/bash` has the extra capabilities because the PSPs won't allow it.

On 2nd thought we *may* be able to achieve what we want with a **patched JVM**, however this would only solve the problem for `IPC_LOCK` since `SYS_RESOURCE` is still needed by `/bin/bash`.

Reply:
thanks @johananl for the confirmation, we will go with this solution for now. I ll merge it in a while after some testing on my side.

Reply:
Great. Please do test it thoroughly as I may have missed something, especially things like unexpected Cassandra behavior in either of the modes ("normal" / "privileged").

Reply:
>On 2nd thought we may be able to achieve what we want with a patched JVM, however this would only solve the problem for IPC_LOCK since SYS_RESOURCE is still needed by /bin/bash.

For the record, I've looked into this and it doesn't seem to be possible because it's not the Java process which runs `ulimit -l unlimited`. Even if we run the `ulimit` command [here](https://github.com/instaclustr/cassandra-operator/tree/superdupertopsecretrewrite/docker/cassandra/default-config/cassandra-env.sh.d) it will still be executed by `/bin/bash`.